src/main/java/com/avispl/symphony/dal/infrastructure/management/philips/huebridge/PhilipsHueDeviceCommunicator.java

```diff
@@ -475,58 +602,1093 @@ public List<AggregatedDevice> retrieveMultipleStatistics() {
+				} else if (PhilipsConstant.MOTION_SENSOR.equalsIgnoreCase(type)) {
+					retrieveDeviceByTypeIsMotionSensor(listMetadataDevice.get(deviceId));
+				}
+				if (logger.isDebugEnabled()) {
+					Long time = System.currentTimeMillis() - startTime;
+					logger.debug(String.format("Finished fetch %s details info in worker thread: %s", aggregatedDeviceList.get(deviceId).getDeviceName(), time));
+				}
+			}
+		} catch (Exception e) {
+			logger.error(String.format("Exception during retrieve '%s' data processing", e.getMessage()), e);
+		}
+	}
+
+	/**
+	 * Retrieve device with type is button
+	 *
+	 * @param aggregatedDeviceResponse instance in AggregatedDeviceResponse DTO
+	 */
+	private void retrieveDeviceByTypeIsButton(AggregatedDeviceResponse aggregatedDeviceResponse) {
+		String deviceID = aggregatedDeviceResponse.getId();
+		String buttonID = PhilipsConstant.EMPTY_STRING;
+		Optional<ServicesResponse> servicesResponseStream = Arrays.stream(aggregatedDeviceResponse.getServices()).filter(item -> item.getType().equalsIgnoreCase(PhilipsConstant.DEVICE_POWER)).findFirst();
+		if (servicesResponseStream.isPresent()) {
+			buttonID = servicesResponseStream.get().getId();
+		}
+		String request = PhilipsURL.BUTTON_POWER.getUrl().concat(buttonID);
+		try {
+			DevicePowerWrapper powerWrapper = doGet(request, DevicePowerWrapper.class);
+			if (powerWrapper != null) {
+				PowerDevice powerDevice = powerWrapper.getData()[0];
+				Map<String, String> oldProperties = aggregatedDeviceList.get(deviceID).getProperties();
+				oldProperties.put(PhilipsConstant.BATTERY_LEVEL, powerDevice.getPowerState().getBatteryLevel());
+			} else {
+				logger.error(String.format("%s button device type is empty", aggregatedDeviceList.get(deviceID).getDeviceName()));
+			}
+		} catch (Exception e) {
+			logger.error(String.format("Error while retrieve %s device detail info: %s ", aggregatedDeviceList.get(deviceID).getDeviceName(), e.getMessage()), e);
+		}
+	}
+
+	/**
+	 * Retrieve device with type is motion sensor
+	 *
+	 * @param aggregatedDeviceResponse instance in AggregatedDeviceResponse DTO
+	 */
+	private void retrieveDeviceByTypeIsMotionSensor(AggregatedDeviceResponse aggregatedDeviceResponse) {
+		String deviceID = aggregatedDeviceResponse.getId();
+		String motionID = PhilipsConstant.EMPTY_STRING;
+		Optional<ServicesResponse> servicesResponseStream = Arrays.stream(aggregatedDeviceResponse.getServices()).filter(item -> item.getType().equalsIgnoreCase(PhilipsConstant.MOTION_SENSOR))
+				.findFirst();
+		if (servicesResponseStream.isPresent()) {
+			motionID = servicesResponseStream.get().getId();
+		}
+		String request = PhilipsURL.MOTION_SENSOR.getUrl().concat(motionID);
+		try {
+			MotionSensorWrapper motionSensorWrapper = doGet(request, MotionSensorWrapper.class);
+			if (motionSensorWrapper != null) {
+				MotionDevice motionDevice = motionSensorWrapper.getData()[0];
+				Map<String, String> oldProperties = aggregatedDeviceList.get(deviceID).getProperties();
```

Comment:
You could extract aggregatedDeviceList.get(deviceID) to a variable and reuse it later.

Reply:
Same comment for  motionDevice.getMotion()